tinygrad/tinygrad
# [wip] onnx bounty

```
tinygrad                       :   2852
tinygrad/codegen               :   1372
tinygrad/engine                :   1201
tinygrad/nn                    :    356
tinygrad/renderer              :    798
tinygrad/runtime               :   2178
tinygrad/runtime/graph         :    302
tinygrad/runtime/onnx          :    503
tinygrad/runtime/support       :    448
tinygrad/shape                 :    347
tinygrad/viz                   :    112

total line count: 10469
```

my attempt
wip... 

=== Things I did ===

tests
- added `test/external/external_model_verify.py` to do deeper validation for our onnx. Useful tool for debugging (like float16 inaccuracy actually starts with our conv)
- `compile2.py` used strictly float32 before so I left it that way
- `test_onnx.py` now uses onnx_model model_proto specified dtypes

onnx_ops
- Use newer tensor apis here and there to simplify code.
- Moved tensor_methods into onnx.py and added `equivalent_tensor_methods` to match differently named onnx ops with our tensor methods. (remove more ops)
- re-did some ops since the crap I wrote before were some of the first code I've ever written
- moved some stuff to tensor
- removed all the to_python_consts in the ops and moved it into one place in onnx.py

tensor
- added padding modes to pad2d since our pad2d matches torch's pad
- added ceil_mode for max_pool2d and avg_pool2d
- added hardsigmoid
- added silu
- added nll_loss

onnx
- improved input parser to turn non-Tensor inputs into Tensor with dtype specified by model_proto and validate Tensor inputs with model_proto. Pins them to proto specifications.
- brought back float16 properly. Have tools to check if onnx run is in accordance with specifications in model_proto
- improved layout


=== TODOs ===

things I'm sure I have todo
- [ ] review my TODOs in my code
- [ ] move more ops into tensor (wip)
- [ ] add to test_onnx.py
- [ ] figure out how to deal with float16. The atol and rtols have to be really high for the test to pass.
- [ ] review external_test_onnx_backend skipped tests. A few more can be enabled
- [ ] update onnx to 1.17.0
- [ ] finally split into smaller PRs and try to get it merged piece by piece if my PR gets locked or something. (wip)

things I'm unsure but have been thinking about
- ~make onnx runner into a class~
- ~remove onnx_ops.py and just have 1 file~
- add asymmetric padding support for convtranspose2d, avg_pool2d and max_pool2d
- maybe add some onnx stuff to process replay to track regressions?
